File: javascript/binary-search-tree/binary-search-tree.js
class BinarySearchTree {

    constructor(data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }

    insert(data) {
        if (data > this.data) {
            if (!this.right) {
                this.right = new BinarySearchTree(data);
            } else {
                this.right.insert(data);
            }
        } else {
            if (!this.left) {
                this.left = new BinarySearchTree(data);
            } else {
                this.left.insert(data);
            }
        }
    }

    each(fn) {
        if (this.left) {
            this.left.each(fn);
        }
        fn(this.data);
        if (this.right) {
            this.right.each(fn);
        }
    }
}



module.exports = BinarySearchTree;

File: fsharp/binary-search-tree/BinarySearchTree.fs
﻿module BinarySearchTree

open System

type BST<'a> =
    | Leaf
    | Node of BST<'a> * 'a * BST<'a>

let private simgleton (value: 'a): BST<'a> = 
    Node (Leaf, value, Leaf)

let left node = 
    match node with
    | Node (Leaf, _, _) -> None
    | Node (x, _, _) -> Some x
    | _ -> failwith "Oops"

let right node = 
    match node with
    | Node (_, _, Leaf) -> None
    | Node (_, _, x) -> Some x
    | _ -> failwith "Oops"

let data node = 
    match node with
    | Node (_, x, _) -> x
    | _ -> failwith "Oops"

let create (items: 'a list): BST<'a> =
    simgleton items.[0]

let sortedData node = failwith "You need to implement this function."